### PR TITLE
Document tags box JavaScript

### LINKED ISSUE #6: Document wp-admin/js/tags-box.js

### PR DESCRIPTION
**Pull request by jcomack**
Props to @atimmer for helping me out with this one.
Props to @CarolineGeven for helping as well.

Fixes #6

### CODE REVIEW

File: src/wp-admin/js/tags-box.js [hunk 1]
```diff
@@ -4,24 +4,60 @@
...
 ( function( $ ) {
+
 	var tagDelimiter = ( window.tagsSuggestL10n && window.tagsSuggestL10n.tagDelimiter ) || ',';
 
-	// Return an array with any duplicate, whitespace or empty values removed
+	/**
+	 * Filters all items from an array into a new array containing only the unique items.
+	 * This also excludes whitespace or empty values.
+	 *
+	 * @summary Filters all items from an array into a new array containing only the unique items.
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: The summary is the same as the first line. Is this desirable?

File: src/wp-admin/js/tags-box.js [hunk 2]
```diff
@@ -4,24 +4,60 @@
...
+	 *
+	 * @since 4.2
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: This should be `4.2.0`

File: src/wp-admin/js/tags-box.js [hunk 3]
```diff
@@ -4,24 +4,60 @@
...
 
 		return out;
 	};
 
+	/**
+	 * The TagBox object.
+	 *
+	 * Contains functions to create and manage tags that can be associated with a post.
+	 *
+	 * @since 4.2
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: This should be `4.2.0`

File: src/wp-admin/js/tags-box.js [hunk 4]
```diff
@@ -4,24 +4,60 @@
...
+	 * @namespace
+	 * @global
+	 */
 	tagBox = {
+
+		/**
+		 * Cleans up tags by converting the tag delimiter to comma's and by removing extra spaces and comma's.
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: comma's should be commas

File: src/wp-admin/js/tags-box.js [hunk 5]
```diff
@@ -46,19 +94,33 @@ var tagBox, array_unique_noempty;
...
 			thetags.val( this.clean( new_tags.join( tagDelimiter ) ) );
 
 			this.quickClicks( taxbox );
 			return false;
 		},
 
+		/**
+		 * Creates all the links, buttons and fields for adding and editing tags.
+		 *
+		 * @summary Creates clickable links, buttons and fields for adding or editing tags.
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: The summary doesn't add something new to the general method description.

File: src/wp-admin/js/tags-box.js [hunk 6]
```diff
@@ -73,6 +135,21 @@ var tagBox, array_unique_noempty;
 			current_tags = thetags.val().split( tagDelimiter );
 			tagchecklist.empty();
 
+			/**
+			 * Creates a new span for every available tag and creates a delete button if tag editing is enabled before
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: This is pretty hard to read. Can you simplify it?

File: src/wp-admin/js/tags-box.js [hunk 7]
```diff
@@ -95,6 +175,19 @@ var tagBox, array_unique_noempty;
 						'<span class="screen-reader-text">' + window.tagsSuggestL10n.removeTerm + ' ' + val + '</span>' +
 						'</button>' );
 
+					/**
+					 * Handles the click and keypress event when an event is called on a tag and reparses the tags
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: Can you rephrase this in two separate sentences?

File: src/wp-admin/js/tags-box.js [hunk 8]
```diff
@@ -95,6 +175,19 @@ var tagBox, array_unique_noempty;
...
+					 * in the tag box.
+					 * Also sets the focus on the new tag field is the enter key is pressed.
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: `if the enter key is pressed`

File: src/wp-admin/js/tags-box.js [hunk 9]
```diff
@@ -95,6 +175,19 @@ var tagBox, array_unique_noempty;
...
+					 *
+					 * @summary Handles the click and keypress event when an event is called on a tag.
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: `when an event` is the same event as the click/keypress event. Now it looks like any random event is triggering something.

File: src/wp-admin/js/tags-box.js [hunk 10]
```diff
@@ -111,17 +204,31 @@ var tagBox, array_unique_noempty;
...
 				// Append the span to the tag list.
 				tagchecklist.append( span );
 			});
+
 			// The buttons list is built now, give feedback to screen reader users.
 			tagBox.screenReadersMessage();
 		},
 
+		/**
+		 * Flushes the tags on save to a hidden field as a comma separated list.
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: Please move `on save` to the end of this line

File: src/wp-admin/js/tags-box.js [hunk 11]
```diff
@@ -111,17 +204,31 @@ var tagBox, array_unique_noempty;
...
+		 *
+		 * @summary Flushes tags on save and on delete.
+		 *
+		 * @since 4.2
+		 *
+		 * @memberOf tagBox
+		 *
+		 * @param {Object} el The container HTML element.
+		 * @param {Object|boolean} a Is either a link from the tag cloud or a hard set boolean value.
+		 * @param {*} f Determines whether or not focus should be applied to the input field.
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: What kind of type is `f` ?

File: src/wp-admin/js/tags-box.js [hunk 12]
```diff
@@ -158,16 +265,50 @@ var tagBox, array_unique_noempty;
 			return false;
 		},
 
+		/**
+		 * Retrieves the available tags from the database and forms a tagcloud.
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: Please change `and forms a tagcloud` to `to form a tagcloud`

File: src/wp-admin/js/tags-box.js [hunk 13]
```diff
@@ -158,16 +265,50 @@ var tagBox, array_unique_noempty;
...
 				r = $( '<p id="tagcloud-' + tax + '" class="the-tagcloud">' + r + '</p>' );
 
+				/**
+				 * Flushes the tagbox whenever an anchor or the tagcloud is clicked.
+				 *
+				 * @summary Flushes the tagbox when clicked.
+				 *
+				 * @since 4.2
+				 *
+				 * @returns {bool} Always returns false.
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: This should be `boolean`

File: src/wp-admin/js/tags-box.js [hunk 14]
```diff
@@ -209,6 +350,16 @@ var tagBox, array_unique_noempty;
 			window.wp.a11y.speak( message, 'assertive' );
 		},
 
+		/**
+		 * Initializes the tags box by setting up the links, buttons and adds event handling.
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: Please change this into: Initializes the tags box by setting up the links, buttons. Also adds event handling.

File: src/wp-admin/js/tags-box.js [hunk 15]
```diff
@@ -221,6 +372,16 @@ var tagBox, array_unique_noempty;
...
 			});
 
+			/**
+			 * Handles the flushing of the tagbox whenever the enter key is released in the new tag field.
+			 *
+			 * @summary Handles the flushing of the tagbox when adding a new tag.
+			 *
+			 * @since 4.2
+			 *
+			 * @param {Event} e The window event to handle.
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: I see `event` is coming in as a parameter.

File: src/wp-admin/js/tags-box.js [hunk 16]
```diff
@@ -237,15 +398,31 @@ var tagBox, array_unique_noempty;
 				$( element ).wpTagsSuggest();
 			});
 
-			// save tags on post save/publish
+			/**
+			 * Flushes tags whenever the post is saved.
+			 *
+			 * @summary Flushes tags whenever the post is saved.
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: This doesn't add anything.

File: src/wp-admin/js/tags-box.js [hunk 17]
```diff
@@ -237,15 +398,31 @@ var tagBox, array_unique_noempty;
...
 			$('#post').submit(function(){
 				$('div.tagsdiv').each( function() {
 					tagBox.flushTags(this, false, 1);
 				});
 			});
 
-			// Fetch and toggle the Tag cloud.
-			$('.tagcloud-link').click(function(){
+			/**
+			 * Unbinds the click event and toggles all sibling tagcloud elements when clicking on the tagcloud-link.
```
Comment (andizer) [inline src/wp-admin/js/tags-box.js]: I would change this into
> When clicking on the tagcloud-link element, unbinds the click event and toggles all sibling tagcloud elements.

**Comment by andizer** [comment]
Forgot to say: `@summary` should be placed above the long description. https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions

File: src/wp-admin/js/tags-box.js [hunk 18]
```diff
@@ -4,24 +4,60 @@
...
+	 * Filters unique items and returns a new array.
+	 *
+	 * Filters all items from an array into a new array containing only the unique items.
+	 * This also excludes whitespace or empty values.
+	 *
+	 * @since 4.2.0
+	 *
+	 * @param {Array} array The array to filter through.
+	 *
+	 * @returns {Array} A new array containing only the unique items.
```
Comment (IreneStr) [inline src/wp-admin/js/tags-box.js]: Please add `@global`

**Comment by CarolineGeven** [comment]
Patch submitted: https://core.trac.wordpress.org/ticket/43499

**Comment by atimmer** [comment]
Has been committed to core in https://core.trac.wordpress.org/changeset/42963/trunk.
